File: app/src/main/java/com/stevesoltys/backup/transport/ConfigurableBackupTransport.java
package com.stevesoltys.backup.transport;

import android.app.backup.BackupTransport;
import android.app.backup.RestoreDescription;
import android.app.backup.RestoreSet;
import android.content.Context;
import android.content.pm.PackageInfo;
import android.net.Uri;
import android.os.ParcelFileDescriptor;
import android.util.Log;

import com.stevesoltys.backup.transport.component.BackupComponent;
import com.stevesoltys.backup.transport.component.RestoreComponent;
import com.stevesoltys.backup.transport.component.provider.ContentProviderBackupComponent;
import com.stevesoltys.backup.transport.component.provider.ContentProviderRestoreComponent;

import static android.app.backup.BackupAgent.FLAG_CLIENT_SIDE_ENCRYPTION_ENABLED;
import static android.os.Build.VERSION.SDK_INT;

/**
 * @author Steve Soltys
 */
public class ConfigurableBackupTransport extends BackupTransport {

    private static final String TRANSPORT_DIRECTORY_NAME =
            "com.stevesoltys.backup.transport.ConfigurableBackupTransport";

    private static final String TAG = TRANSPORT_DIRECTORY_NAME;

    private final BackupComponent backupComponent;

    private final RestoreComponent restoreComponent;

    ConfigurableBackupTransport(Context context) {
        backupComponent = new ContentProviderBackupComponent(context);
        restoreComponent = new ContentProviderRestoreComponent(context);
    }

    public void prepareRestore(String password, Uri fileUri) {
        restoreComponent.prepareRestore(password, fileUri);
    }

    @Override
    public String transportDirName() {
        return TRANSPORT_DIRECTORY_NAME;
    }

    @Override
    public String name() {
        // TODO: Make this class non-static in ConfigurableBackupTransportService and use Context and a ComponentName.
        return this.getClass().getName();
    }

    @Override
    public int getTransportFlags() {
        if (SDK_INT >= 28) return FLAG_CLIENT_SIDE_ENCRYPTION_ENABLED;
        return 0;
    }

    @Override
    public boolean isAppEligibleForBackup(PackageInfo targetPackage, boolean isFullBackup) {
        return true;
    }

    @Override
    public long requestBackupTime() {
        return backupComponent.requestBackupTime();
    }

    @Override
    public String dataManagementLabel() {
        return backupComponent.dataManagementLabel();
    }

    @Override
    public int initializeDevice() {
        return backupComponent.initializeDevice();
    }

    @Override
    public String currentDestinationString() {
        return backupComponent.currentDestinationString();
    }

    /* Methods related to Backup */

    @Override
    public int performBackup(PackageInfo packageInfo, ParcelFileDescriptor inFd, int flags) {
        return backupComponent.performIncrementalBackup(packageInfo, inFd, flags);
    }

    @Override
    public int performBackup(PackageInfo targetPackage, ParcelFileDescriptor fileDescriptor) {
        Log.w(TAG, "Warning: Legacy performBackup() method called.");
        return performBackup(targetPackage, fileDescriptor, 0);
    }

    @Override
    public int checkFullBackupSize(long size) {
        return backupComponent.checkFullBackupSize(size);
    }

    @Override
    public int performFullBackup(PackageInfo targetPackage, ParcelFileDescriptor socket, int flags) {
        // TODO handle flags
        return performFullBackup(targetPackage, socket);
    }

    @Override
    public int performFullBackup(PackageInfo targetPackage, ParcelFileDescriptor fileDescriptor) {
        return backupComponent.performFullBackup(targetPackage, fileDescriptor);
    }

    @Override
    public int sendBackupData(int numBytes) {
        return backupComponent.sendBackupData(numBytes);
    }

    @Override
    public void cancelFullBackup() {
        backupComponent.cancelFullBackup();
    }

    @Override
    public int finishBackup() {
        return backupComponent.finishBackup();
    }

    @Override
    public long requestFullBackupTime() {
        return backupComponent.requestFullBackupTime();
    }

    @Override
    public long getBackupQuota(String packageName, boolean isFullBackup) {
        return backupComponent.getBackupQuota(packageName, isFullBackup);
    }

    @Override
    public int clearBackupData(PackageInfo packageInfo) {
        return backupComponent.clearBackupData(packageInfo);
    }

    public void backupFinished() {
        backupComponent.backupFinished();
    }

    /* Methods related to Restore */

    @Override
    public long getCurrentRestoreSet() {
        return restoreComponent.getCurrentRestoreSet();
    }

    @Override
    public int startRestore(long token, PackageInfo[] packages) {
        return restoreComponent.startRestore(token, packages);
    }

    @Override
    public int getNextFullRestoreDataChunk(ParcelFileDescriptor socket) {
        return restoreComponent.getNextFullRestoreDataChunk(socket);
    }

    @Override
    public RestoreSet[] getAvailableRestoreSets() {
        return restoreComponent.getAvailableRestoreSets();
    }

    @Override
    public RestoreDescription nextRestorePackage() {
        return restoreComponent.nextRestorePackage();
    }

    @Override
    public int getRestoreData(ParcelFileDescriptor outputFileDescriptor) {
        return restoreComponent.getRestoreData(outputFileDescriptor);
    }

    @Override
    public int abortFullRestore() {
        return restoreComponent.abortFullRestore();
    }

    @Override
    public void finishRestore() {
        restoreComponent.finishRestore();
    }
}

File: app/src/main/java/com/stevesoltys/backup/transport/ConfigurableBackupTransport.kt
package com.stevesoltys.backup.transport

import android.app.backup.BackupAgent.FLAG_CLIENT_SIDE_ENCRYPTION_ENABLED
import android.app.backup.BackupTransport
import android.app.backup.RestoreDescription
import android.app.backup.RestoreSet
import android.content.Context
import android.content.Intent
import android.content.pm.PackageInfo
import android.os.ParcelFileDescriptor
import android.util.Log
import com.stevesoltys.backup.settings.SettingsActivity

val TRANSPORT_ID: String = ConfigurableBackupTransport::class.java.name

private const val TRANSPORT_DIRECTORY_NAME = "com.stevesoltys.backup.transport.ConfigurableBackupTransport"
private val TAG = ConfigurableBackupTransport::class.java.simpleName

/**
 * @author Steve Soltys
 * @author Torsten Grote
 */
class ConfigurableBackupTransport internal constructor(private val context: Context) : BackupTransport() {

    private val pluginManager = PluginManager(context)
    private val backupCoordinator = pluginManager.backupCoordinator
    private val restoreCoordinator = pluginManager.restoreCoordinator

    override fun transportDirName(): String {
        return TRANSPORT_DIRECTORY_NAME
    }

    override fun name(): String {
        return TRANSPORT_ID
    }

    override fun getTransportFlags(): Int {
        return FLAG_CLIENT_SIDE_ENCRYPTION_ENABLED
    }

    override fun dataManagementIntent(): Intent {
        return Intent(context, SettingsActivity::class.java)
    }

    override fun dataManagementLabel(): String {
        return "Please file a bug if you see this! 1"
    }

    override fun currentDestinationString(): String {
        return "Please file a bug if you see this! 2"
    }

    // ------------------------------------------------------------------------------------
    // General backup methods
    //

    override fun initializeDevice(): Int {
        return backupCoordinator.initializeDevice()
    }

    override fun isAppEligibleForBackup(targetPackage: PackageInfo, isFullBackup: Boolean): Boolean {
        return backupCoordinator.isAppEligibleForBackup(targetPackage, isFullBackup)
    }

    override fun getBackupQuota(packageName: String, isFullBackup: Boolean): Long {
        return backupCoordinator.getBackupQuota(packageName, isFullBackup)
    }

    override fun clearBackupData(packageInfo: PackageInfo): Int {
        return backupCoordinator.clearBackupData(packageInfo)
    }

    override fun finishBackup(): Int {
        return backupCoordinator.finishBackup()
    }

    // ------------------------------------------------------------------------------------
    // Key/value incremental backup support
    //

    override fun requestBackupTime(): Long {
        return backupCoordinator.requestBackupTime()
    }

    override fun performBackup(packageInfo: PackageInfo, inFd: ParcelFileDescriptor, flags: Int): Int {
        return backupCoordinator.performIncrementalBackup(packageInfo, inFd, flags)
    }

    override fun performBackup(targetPackage: PackageInfo, fileDescriptor: ParcelFileDescriptor): Int {
        Log.w(TAG, "Warning: Legacy performBackup() method called.")
        return performBackup(targetPackage, fileDescriptor, 0)
    }

    // ------------------------------------------------------------------------------------
    // Full backup
    //

    override fun requestFullBackupTime(): Long {
        return backupCoordinator.requestFullBackupTime()
    }

    override fun checkFullBackupSize(size: Long): Int {
        return backupCoordinator.checkFullBackupSize(size)
    }

    override fun performFullBackup(targetPackage: PackageInfo, socket: ParcelFileDescriptor, flags: Int): Int {
        return backupCoordinator.performFullBackup(targetPackage, socket, flags)
    }

    override fun performFullBackup(targetPackage: PackageInfo, fileDescriptor: ParcelFileDescriptor): Int {
        Log.w(TAG, "Warning: Legacy performFullBackup() method called.")
        return backupCoordinator.performFullBackup(targetPackage, fileDescriptor, 0)
    }

    override fun sendBackupData(numBytes: Int): Int {
        return backupCoordinator.sendBackupData(numBytes)
    }

    override fun cancelFullBackup() {
        backupCoordinator.cancelFullBackup()
    }

    // ------------------------------------------------------------------------------------
    // Restore
    //

    override fun getAvailableRestoreSets(): Array<RestoreSet>? {
        return restoreCoordinator.getAvailableRestoreSets()
    }

    override fun getCurrentRestoreSet(): Long {
        return restoreCoordinator.getCurrentRestoreSet()
    }

    override fun startRestore(token: Long, packages: Array<PackageInfo>): Int {
        return restoreCoordinator.startRestore(token, packages)
    }

    override fun getNextFullRestoreDataChunk(socket: ParcelFileDescriptor): Int {
        return restoreCoordinator.getNextFullRestoreDataChunk(socket)
    }

    override fun nextRestorePackage(): RestoreDescription? {
        return restoreCoordinator.nextRestorePackage()
    }

    override fun getRestoreData(outputFileDescriptor: ParcelFileDescriptor): Int {
        return restoreCoordinator.getRestoreData(outputFileDescriptor)
    }

    override fun abortFullRestore(): Int {
        return restoreCoordinator.abortFullRestore()
    }

    override fun finishRestore() {
        restoreCoordinator.finishRestore()
    }

}
